File: corpus/typescript-efficiency-design-v2/changed/enrichment.ts
interface Item { active: boolean; owner?: { enabled: boolean } }
interface Repository {
  load(id: string): Promise<Item>;
  loadMany(ids: string[]): Promise<Item[]>;
}

class ResultArrayFactory {
  static create(): Item[] { return []; }
}

export async function enrich(ids: string[], repo: Repository): Promise<Item[]> {
  const results = ResultArrayFactory.create();
  for (const id of ids) {
    const item = await repo.load(id);
    if (item.active) {
      if (item.owner) {
        if (item.owner.enabled) results.push(item);
      }
    }
  }
  return results;
}

File: corpus/dart-efficiency-design-v2/base/lib/enrichment.dart
class Owner { final bool enabled; Owner(this.enabled); }
class Item { final bool active; final Owner? owner; Item(this.active, this.owner); }
abstract class Repository {
  Future<Item> load(String id);
  Future<List<Item>> loadMany(List<String> ids);
}

Future<List<Item>> enrich(List<String> ids, Repository repo) async {
  final results = <Item>[];
  for (final item in await repo.loadMany(ids)) {
    if (item.active && (item.owner?.enabled ?? false)) results.add(item);
  }
  return results;
}
